pkg/login/social/generic_oauth.go

```diff
@@ -294,6 +294,11 @@ func (s *SocialGenericOAuth) extractToken(data *UserInfoJson, token *oauth2.Toke
 		return []byte{}, false
 	}
 
+	if role := s.extractRole(data, payload); s.roleAttributePath != "" && role == "" {
```

Comment:
This change doesn't make sense to me. What if you want to extract email address from token and above email check succeeds, but then this extraction of role doesn't find any role. Result will be that you will maybe get a role from UserInfo, but no email.

For now it's either extract from token or UserInfo, not both. Quite big refactoring is needed to support this so would suggest open a feature request for this specific use case and handle that change in a separate PR.

Reply:
@marefr I'll see if I can put together a test case for this so we can ensure we don't have a regression on email flow. I think without this specific line change we could definitely get an improvement but I agree, there's a good chunk of refactoring to be done.

Reply:
Okay, let me know.when you have something ready or want feedback. 

One idea for refactoring/simplify this may be to store token payload bytes on the UserInfoJson struct. 

Above change probably needs something along the lines:
 - If email or role are not found in token, user info should be inspected
 - if both email and role are found in token user info should not be inspected

Reply:
Hi @seanson  and @marefr !

I have looked into this and prepared a branch based on this one that picks up @marefr 's suggestions and should complete the fix. It is not a trivial fix so I can see how @seanson  might have gotten stuck on completing this. I tried to stick to the objective, though, and keep the fix as small as possbile.

I am not sure how to go from here. @seanson I think it would be easiest if you simply added my commits to this PR if you agree with the fix. Otherwise I could also open a new one. Please let me know what you think, I'd really like to see this issue fixed.

Thanks. ;-)

https://github.com/henninge/grafana/commits/17149-fix-oauth-data

Reply:
@henninge - Thanks for the help! Always glad for a helping hand. I'll merge this commit into this PR and write a test case to cover @marefr's flow.

Reply:
Changes rebased against master with the main auth cherry-picked and my refactored tests added. Should be good for review now. @henninge please ensure you've signed the CLA! :)

Reply:
Cool! Did you notice that there were two commits? I had refactored the tests, too. I think I came up with a couple extra test cases.

Reply:
@henninge - I had missed the extra test cases as we had both added the email/role checks. I've merged the new tests in with the existing test cleanup in the new commit.

I believe this should be good for review now.